### PR TITLE
Remove support for legacy driver and upgrade to 0.4.1 in ClickHouse connector

### LINKED ISSUE #16188: upgrade clickhouse jdbc driver to the latest version
Was talking to clickhouse developers regarding some issues I had and it turned out it's the jdbc driver ,
Trino is using clickhouse jdbc driver 0.3.2-patch3 , and alot of enhancements and bug fixing has been added to the driver since then, that would be great if we could upgrade the driver.

Or least i think it would be great to get the latest patch (0.3.2-patch11) , if version 0.4 is a breaking version.

https://github.com/ClickHouse/clickhouse-java/releases

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

* Remove support for legacy driver in ClickHouse since it took more than 1 year after becoming the deprecation
* Upgrade ClickHouse JDBC driver to 0.4.1. Even the latest driver doesn't provide `ResultSetMetaData` for `query` pass-through table function

Fixes #16188

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# ClickHouse
* Remove support for legacy JDBC driver. ({issue}`16193`)
```

### CODE REVIEW

**Comment by zhicwu** [comment]
Hi @ebyhr, can we bump to 0.4.1? Although it's just out for 2 days, but it fixed a few critical bugs found in 0.4.0 like 1) incorrect nested array value; and 2) broken serde for `Nested` data type which may lead to failures when deserializing `Nested` and `JSON` values.

Apart from that, I think there's one breaking change worthy of mention here. Starting from 0.4.0, to support binary string(mapping `String` and `FixedString` to byte array), you have to explicitly set `use_binary_string` to `true`. This is because 1) ~10% performance improvement for deserializing each String column and less memory footprint when use_binary_string=false, which is default; and 2) binary data can be also represented as `Array(*Int8)` in ClickHouse.

**Comment by ebyhr** [comment]
@zhicwu Thanks for sharing the information! Changed to 0.4.1 and enabled `use_binary_string` property.

**Comment by ebyhr** [comment]
`web-ui-checks` hit #16179

**Comment by zhicwu** [comment]
> @zhicwu Thanks for sharing the information! Changed to 0.4.1 and enabled `use_binary_string` property.

Thank you and apologize for additional changes at your end. Does it make sense to you to map `VARBINARY` to  `Array(UInt8)` in ClickHouse?

**Comment by ebyhr** [comment]
@zhicwu  What about read logic? Reading `Array(UInt8)` as `VARBINARY` looks weird to me.

**Comment by zhicwu** [comment]
> @zhicwu What about read logic? Reading `Array(UInt8)` as `VARBINARY` looks weird to me.

I'm not joking, but is `Array(BYTE)` better? `BYTE` is an alias of `Int8`.

Looks like the mapped type in ClickHouse doesn't matter when reading/writing varbinary in trino, according to code below:

https://github.com/trinodb/trino/blob/fba39d69e972b2d7fd8673ff2e46be212d1550ea/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java#L348-L356

**Comment by ebyhr** [comment]
@zhicwu Mapping to `Array(UInt8)` or `Array(BYTE)` changes the behavior of `NULL` handling as far as I confirmed. I will keep as-is in this PR. Please feel free to send a new one after this PR.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClientModule.java [hunk 1]
```diff
@@ -52,16 +50,8 @@ public void configure(Binder binder)
...
-    }
-
-    private static Driver createDriver(ClickHouseConfig config)
+    public static ConnectionFactory createConnectionFactory(BaseJdbcConfig config, CredentialProvider credentialProvider)
     {
-        if (config.isLegacyDriver()) {
-            return new ru.yandex.clickhouse.ClickHouseDriver();
-        }
-        return new com.clickhouse.jdbc.ClickHouseDriver();
+        return new ClickHouseConnectionFactory(new DriverConnectionFactory(new com.clickhouse.jdbc.ClickHouseDriver(), config, credentialProvider));
```
Comment (Praveen2112) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClientModule.java]: nit: Now we can directly import `ClickHouseDriver` right ?

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java [hunk 2]
```diff
@@ -67,20 +63,11 @@ private static ClickHouseContainer createContainer(DockerImageName image)
             @Override
             public String getDriverClassName()
             {
-                return getClickhouseDriverClassName(image);
+                return CLICKHOUSE_LATEST_DRIVER_CLASS_NAME;
             }
```
Comment (Praveen2112) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java:67]: Do we need to override this `getDriverClassName` - won't `ClickHouseContainer` maps to the latest one.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 3]
```diff
@@ -179,7 +180,6 @@
 import static java.time.ZoneOffset.UTC;
 import static java.util.Locale.ENGLISH;
 import static java.util.Objects.requireNonNull;
-import static ru.yandex.clickhouse.ClickHouseUtil.escape;
```
Comment (Praveen2112) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:182]: Shouldn't this be a part of previous commit or is it only when we update 0.4.1 version.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClientModule.java [hunk 4]
```diff
@@ -52,6 +56,8 @@ public void configure(Binder binder)
     @ForBaseJdbc
     public static ConnectionFactory createConnectionFactory(BaseJdbcConfig config, CredentialProvider credentialProvider)
     {
-        return new ClickHouseConnectionFactory(new DriverConnectionFactory(new com.clickhouse.jdbc.ClickHouseDriver(), config, credentialProvider));
+        Properties properties = new Properties();
+        properties.setProperty(USE_BINARY_STRING.getKey(), "true");
```
Comment (Praveen2112) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClientModule.java:60]: Can we add a comment on why we need to set this property.

**Comment by ebyhr** [comment]
Addressed comments.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 5]
```diff
@@ -95,6 +95,7 @@
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BiFunction;
 
+import static com.clickhouse.client.ClickHouseValues.convertToQuotedString;
```
Comment (Praveen2112) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java]: Do we have this API in the legacy version ?

Comment (Praveen2112) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java]: In this case can we squash with the next commit - I'm happy with both the cases.

**Review by Praveen2112 — APPROVED** [review]

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java (hunk 5 above)
Comment (ebyhr) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java]: It existed, but the package was changed in the new version.

**Pull request merged**